src/main/java/org/folio/service/budget/BudgetService.java

```diff
@@ -237,7 +175,91 @@ public Future<Void> closeBudgets(LedgerFiscalYearRollover rollover, DBConn conn)
         getFullTableName(conn.getTenantId(), BUDGET_TABLE), getFullTableName(conn.getTenantId(), FUND_TABLE),
         rollover.getLedgerId(), rollover.getFromFiscalYearId());
 
-    return budgetDAO.updateBatchBudgets(sql, conn)
+    return budgetDAO.updateBatchBudgetsBySql(sql, conn)
+      .mapEmpty();
+  }
+
+  public Future<List<Budget>> getBudgetsByFiscalYearIdsAndFundIdsForUpdate(
+      Map<String, Set<String>> fiscalYearIdToFundIds, DBConn conn) {
+    StringBuilder sb = new StringBuilder();
+    sb.append("SELECT jsonb FROM ")
+      .append(getFullTableName(conn.getTenantId(), BUDGET_TABLE))
+      .append(" WHERE ");
+    Iterator<String> fiscalYearIterator = fiscalYearIdToFundIds.keySet().iterator();
+    while (fiscalYearIterator.hasNext()) {
+      String fiscalYearId = fiscalYearIterator.next();
+      Set<String> fundIds = fiscalYearIdToFundIds.get(fiscalYearId);
+      sb.append("(jsonb->>'fiscalYearId' = '")
+        .append(fiscalYearId)
+        .append("' AND (");
+      Iterator<String> fundIterator = fundIds.iterator();
+      while (fundIterator.hasNext()) {
+        String fundId = fundIterator.next();
+        sb.append("jsonb->>'fundId' = '")
+          .append(fundId)
+          .append("'");
+        if (fundIterator.hasNext()) {
+          sb.append(" OR ");
+        }
+      }
+      sb.append("))");
+      if (fiscalYearIterator.hasNext()) {
+        sb.append(" OR ");
+      }
+    }
+    sb.append(" FOR UPDATE");
+    String sql = sb.toString();
+    return budgetDAO.getBudgetsBySql(sql, Tuple.tuple(), conn);
+  }
+
+  private Future<Void> deleteAllocationTransactions(Budget budget, DBConn conn) {
+    logger.debug("deleteAllocationTransactions:: Trying to delete allocation transactions with fund id {}", budget.getFundId());
+
+    String sql ="DELETE FROM "+ getFullTableName(conn.getTenantId(), TRANSACTIONS_TABLE)
+      + " WHERE  (jsonb->>'fromFundId' = '"+ budget.getFundId() +"' OR jsonb->>'toFundId' = '"+ budget.getFundId() + "')"
+      + " AND (jsonb->>'fiscalYearId')::text = '" + budget.getFiscalYearId() + "' AND (jsonb->>'transactionType')::text = 'Allocation'";
```

Comment:
I will change all that.

Reply:
thanks, now performance would be even more that we thought intially, expessionally for transactions, because there are lot of closed transactions from previous fiscal years and our queries will use indexes now